lib/Parser/Parse.cpp

```diff
@@ -8850,36 +8853,42 @@ ParseNodePtr Parser::ParseExpr(int oplMin,
 
             if (buildAST)
             {
-                pnode = CreateBinNode(nop, pnode, pnodeT);
-                Assert(pnode->AsParseNodeBin()->pnode2 != NULL);
-                if (pnode->AsParseNodeBin()->pnode2->nop == knopFncDecl)
+                Assert(pnode2 != nullptr);
+                if (pnode2->nop == knopFncDecl)
                 {
                     Assert(hintLength >= hintOffset);
-                    pnode->AsParseNodeBin()->pnode2->AsParseNodeFnc()->hint = pNameHint;
-                    pnode->AsParseNodeBin()->pnode2->AsParseNodeFnc()->hintLength = hintLength;
-                    pnode->AsParseNodeBin()->pnode2->AsParseNodeFnc()->hintOffset = hintOffset;
+                    pnode2->AsParseNodeFnc()->hint = pNameHint;
+                    pnode2->AsParseNodeFnc()->hintLength = hintLength;
+                    pnode2->AsParseNodeFnc()->hintOffset = hintOffset;
 
-                    if (pnode->AsParseNodeBin()->pnode1->nop == knopDot)
+                    if (pnode1->nop == knopDot)
                     {
-                        pnode->AsParseNodeBin()->pnode2->AsParseNodeFnc()->isNameIdentifierRef = false;
+                        pnode2->AsParseNodeFnc()->isNameIdentifierRef = false;
                     }
-                    else if (pnode->AsParseNodeBin()->pnode1->nop == knopName)
+                    else if (pnode1->nop == knopName)
                     {
-                        PidRefStack *pidRef = pnode->AsParseNodeBin()->pnode1->AsParseNodeName()->pid->GetTopRef();
+                        PidRefStack *pidRef = pnode1->AsParseNodeName()->pid->GetTopRef();
                         pidRef->isFuncAssignment = true;
                     }
                 }
-                if (pnode->AsParseNodeBin()->pnode2->nop == knopClassDecl && pnode->AsParseNodeBin()->pnode1->nop == knopDot)
+                else if (pnode2->nop == knopClassDecl && pnode1->nop == knopDot)
                 {
-                    Assert(pnode->AsParseNodeBin()->pnode2->AsParseNodeClass()->pnodeConstructor);
+                    Assert(pnode2->AsParseNodeClass()->pnodeConstructor);
 
-                    if (!pnode->AsParseNodeBin()->pnode2->AsParseNodeClass()->pnodeConstructor->pid)
+                    if (!pnode2->AsParseNodeClass()->pnodeConstructor->pid)
                     {
-                        pnode->AsParseNodeBin()->pnode2->AsParseNodeClass()->pnodeConstructor->isNameIdentifierRef = false;
+                        pnode2->AsParseNodeClass()->pnodeConstructor->isNameIdentifierRef = false;
                     }
                 }
+                else if (pnode1->nop == knopName && nop == knopIn)
```

Comment:
This is the optimization (the rest is beautification).

Reply:
Looked at it with Jimmy. The current optimization for load already kicks in for patterns like:

var x = "foo";
var obj = {};
if (x in obj) console.log(obj[x]);

So the scenarios that would benefit from this fix seem to be "use a kind of tagging property on objects for which only care about its existence but not the value". Is it still worth taking the fix?

---
In reply to: [187660351](https://github.com/Microsoft/ChakraCore/pull/5135#discussion_r187660351) [](ancestors = 187660351)

Reply:
I've also written code that looks like `if ("foo" in x) use(x.foo)`. Would this optimization help that case, where the property access following the `in` is using the dot operator?

---
In reply to: [187666992](https://github.com/Microsoft/ChakraCore/pull/5135#discussion_r187666992) [](ancestors = 187666992,187660351)

Reply:
if ("foo" in x) won't trigger the new optimization (literal isn't a name) and use(x.foo) is already causing foo to be a property.